Standalone - factor out boot steps from index.php

ufundo opened the pull request:
Overview
----------------------------------------

I've attempted to:

1. move as much as possible out of index.php into core classes (in `\Civi\Standalone` )
2. factor out the bootstrapping that can be shared between index.php and cv, 
3. simplify directory structure and settings interface for Standalone

I've moved 3 to a separate PR to try to make the whole thing a bit more intelligible: https://github.com/civicrm/civicrm-core/pull/30132 

Changes:
- the error handler and invoke functions from index.php are moved to classes in \Civi\Standalone
- the web install UI function is moved to \Civi\Setup\UI

 ~new PathLoader class is added for path configuration in Standalone, aiming for simpler/flatter config using env vars / nicer default~
 ~~amend the standalone installer to use this class to determine paths during the install process~~

- move the installation flag file to `civicrm.standalone.php` and use this as much as possible for the bootstrapping that can be shared between `index.php` and `cv`, but can't live in core (in my head this is: autoloader, classloader, pathloader, settings loader).
- accompanying patch for cv that finds this https://github.com/civicrm/cv/pull/193/commits

~~allows for a settings directory in place of `civicrm.settings.php` - most of the content of the existing template is split out into separate static files (which are bundled statically in the installer), and only those things set by the installer are templated. this also provides a lot of potential/flexibility for packaging civi with different settings, and or smoothing the workflow for merging updates to settings.php with existing installs (you can diverge from master with specific files, and keep others updated with any released updates / you can share files between different installs ... etc)~~
~~allow prepopulating installer database credentials from env variables~~

Ideally it would test with `distmaker` but I haven't got that running, so to simulate the process I have been patching the latest nightly tarball as follows:

- download and extract the tarball
- remove the old flag file 
- apply this patch in the `web/core` directory 
- copy `web/core/setup/res/index.php.txt` to `web/index.php`
- copy `web/core/setup/res/civicrm.standalone.php.txt` to `web/civicrm.standalone.php`
- point webserver at web and run the web installer

Before
----------------------------------------
There's some important un-upgradeable logic in index.php

CV has to do some of the same stuff in parallel - so might end up diverging in what settings it has.

There are quite a few `findThisFile` functions in index.php, which try a few different hard coded options but fail if you do anything else

After
----------------------------------------
As much logic as possible is in the core repo code.

There is a common code path for index.php and cv to load the autoloader / classloader and settings

I've replaced the `findThisFile` functions with explicit variables, based on knowing the relative paths in on our distribution format. I think this makes it much clear and easier to adapt if you do what to diverge from the default distribution.

ufundo commented:
A thought - maybe this doesn't make it into core but could be used for a docker-image-specific `civicrm.settings.php` - where I think the env file config is particularly pertinent... :thinking:

ufundo commented:
So here I added an `if Standalone` block to the settings template.... but what about totally greenfielding it? 🤔 

i.e. add a separate template for Standalone, and trying to strip it right down, removing stuff that isnt relevant or intended for backwards-compatibility 

Disadvantages :
- two templates to maintain going forward
- harder to upgrade from non-Standalone to Standalone (b/c less obvious how to merge any existing civicrm.settings.php into the new template)

Advantages:
- could be a lot cleaner?

@totten whilst I'm on this train.. I think we talked about how it could be preferable to have a directory for settings files? Which could allow for splitting concerns/clearer logic of what does what (ie by default a file for path settings, a file for performance tweaks, a file for crypto keys... ). Generally more navigable. And powerful for sysadmins too I'd think (e.g. you could have a common file for e.g. file path structures / mandatory settings across many deployments, but individual files for crypto keys). Am I getting carried away?

colemanw commented:
➕ for greenfield!

totten commented:
Aside: There's a pending project to change the test-runs. A key distinction is that it will spot-check more variations of UF and PHP version. It's not enabled  by default yet, but I fired a manual run.

On the status for the last commit, you should see a link to [this output](https://test.civicrm.org/duderino/batch/20865809). The `standalone-clean` job shows a `cv` error during installation. (*Haven't looked close - but I thought maybe the stacktrace would be useful.*)

ufundo commented on setup/plugins/blocks/database.civi-setup.php:
```diff
@@ -22,7 +22,15 @@
+      'server' => $dbHost . ':' . $dbPort,
+      'username' => getenv('CIVICRM_DB_USER') ?: '',
+      'database' => getenv('CIVICRM_DB_NAME') ?: '',
+      // should we expose the env var through the web installer ui? 'password' => getenv('CIVICRM_DB_PASS') ?: '',
```
:+1: 

the concern I have with that is that I was hoping index.php would work on the basis of "if site isn't installed, show the installer" (a la Wordpress/Drupal). 

however if the database credentials are provided by ENV vars, but the database goes away in such a way that index.php thinks "oh, the site isn't installed, let's run the installer" then your database credentials are going to then be exposed to the world for all to see.

I was thinking a good solution would be something that checks db credentials provided in the environment, and then if they are good, just confirms to the user what the database name / user are, without exposing the password.
 
all that said... I'm going to try to split the database changes out to a separate PR, as they are independent, and there's going to be enough in here to chew on...

ufundo commented on setup/src/Setup/StandalonePathSettings.php:
```diff
@@ -0,0 +1,353 @@
+      throw new \ValueError('Cannot set empty string path for key: ' . $key);
+    }
+
+    $this->configuredPaths[$key] = $this->configuredPaths[$key] ?? [];
```
you learn something every day!

ufundo commented on setup/src/Setup/StandalonePathSettings.php:
```diff
@@ -0,0 +1,353 @@
+  protected function pathFromDefault($key): ?string
+  {
+    if ($key === 'project_root') {
+      $path = $_SERVER['SERVER_ROOT'] ?? $_SERVER['DOCUMENT_ROOT'] ?? null;
```
haha wishful thinking on my part. this bit needs a bit of cleaning up anyway

ufundo committed:
factor out Standalone bootstrap from index.php

ufundo commented:
retest this please

ufundo commented:
To note, I think the passing tests above more or less completely ignore the changes in this PR.

Actually testing would require [this accompanying patch for cv](https://github.com/civicrm/cv/pull/193/commits)  - otherwise cv won't follow the new boot path, and so buildkit will fail.

ufundo committed:
adapt existing settings file template for standalone

ufundo committed:
dont warn if no settings file

ufundo committed:
update installer for data/public data/private structure

ufundo commented:
> Fatal error: Uncaught RuntimeException: Failed to identify standalone root. (TIP: Set extras.standaloneRoot) in /var/www/standalone.localhost/web/core/setup/plugins/init/Standalone.civi-setup.php on line 76

Acht sorry! Realising I had some old directory structure kicking around when rebasing :/ Should be fixed with additional commits.

ufundo commented:
> Comparing to the upstream main branch; calling scaffold at ➊ requires passing PROJROOT not DOCROOT (where DOCROOT=$PROJROOT/web) and then the installer works. Passing PROJROOT on this PR causes civicrm.standalone.php, data/ and index.php to all be written to the PROJROOT which is not right.


My intention following these changes was $PROJROOT =  $DOCROOT, such that the structure would be:

- PROJROOT = DOCROOT = `/var/www/html/web`
- CORE = `/var/www/html/web/core`
- DATA = `/var/www/html/web/data`
- INDEX = `/var/www/html/web/index.php`

I thought that was the conclusion we came to as simplest default structure for the tarball, which should be the standard?

ufundo commented:
> I think there should be an if file_exists() around that include

Yes there should, added. (This gets more tightly wrapped in the follow up PR)

ufundo commented:
Hurrah! :tada:  

Thanks for bearing with me :pray: 

Will fix the dropped comma that's crept in...

ufundo commented on Civi/Standalone/WebEntrypoint.php:
```diff
@@ -0,0 +1,110 @@
+      return !!CIVICRM_INSTALLED;
+    }
+    if (defined('CIVICRM_DSN')) {
+      // can we autodetect a good value here?
+    }
```
So the reason this is here: [in the follow up PR](https://github.com/civicrm/civicrm-core/pull/30132), the idea is you *can* provide a settings file (and/or env vars) prior to running the installer, with DSN settings in them.

At this point the signal is not whether there is a settings file, but whether the CIVICRM_INSTALLED is set to TRUE in the settings. (The installer writes an additional settings file into the settings directory which adds this line).

To my mind it would be nice if CIVICRM_INSTALLED isn't defined to be able to fallback to checking any provided DSN details to see if they point to a valid install, and not show the installer in this case. But I was struggling to find an existing way to do this at this point in the bootstrap.

ufundo replied:
I can leave out until it's actually functioning though, sure.

ufundo replied:
(commented out)

totten replied:
1. The PR has comments about security / flaky connections / re-installation -- that's a really good point.

1. I relate to the idea of checking DSN -- but ultimately agree that we should *not* check it here/now.
    * There's a scope mismatch. On one side, you have a simple boolean (`WebEntrypoint::checkCiviInstalled()`). On the other side, "installation" refers to a process with various resources, phases, failure-points, future-features. Those are not the same size/complexity. Intuitively, there will be cases where the boolean is an anemic representation of the true installation-status.
    * Aside: There exists `$setup->checkInstalled()`. This actually returns an object which can hold more details. But you're right to *not* use that here. It does more work, and (for 99.9% of pageloads) you don't need that work.
    * There are hypotheticals for more complex things, but I don't think they're important here/now, and I vote to keep the check simple+predictable. (And we're not locked in -- the whole point of this file is to make it easier to change/evolve.)

1. On a gut level, `CIVICRM_INSTALLED` doesn't feel like a configuration value. It kinda feels like repeating the mistakes of  `set_include_path()` and `CRM_Core_ClassLoader::singleton()->register();`. (They're oddballs in the settings file which aren't proper configuration data -- but once you've propagated the template to some large subset of sites, it's PITA to cleanup.*)
    * In other UF's, if you wanted to identify whether you had active/loaded settings, you could check `CIVICRM_SETTINGS_PATH`. (*This value is set right before loading the file; eg [d7](https://github.com/civicrm/civicrm-drupal/blob/7.x-master/civicrm.module#L186-L188), [d8+](https://github.com/civicrm/civicrm-drupal-8/blob/c3988a297bbe852f23a682c66472480dcb549320/src/Civicrm.php#L40-L45), [joomla](https://github.com/civicrm/civicrm-joomla/blob/6287dd5d3614b2f5cedf9081126aa363ad96d4e9/site/civicrm.php#L10-L11).*)
    * I think this would keep the settings-file tidier (*lower maintenance*) and align with other UFs: https://github.com/civicrm/civicrm-core/commit/c309e697f49ede3452dd0bd02fb36cd04c27c210

ufundo replied:
> In other UF's, if you wanted to identify whether you had active/loaded settings, you could check CIVICRM_SETTINGS_PATH. (This value is set right before loading the file; eg [d7](https://github.com/civicrm/civicrm-drupal/blob/7.x-master/civicrm.module#L186-L188), [d8+](https://github.com/civicrm/civicrm-drupal-8/blob/c3988a297bbe852f23a682c66472480dcb549320/src/Civicrm.php#L40-L45), [joomla](https://github.com/civicrm/civicrm-joomla/blob/6287dd5d3614b2f5cedf9081126aa363ad96d4e9/site/civicrm.php#L10-L11).)

So I was deliberately trying to avoid checking CIVICRM_SETTINGS_PATH. If a boolean is an anemic representation of installation status, then the settings path is a bad proxy for the boolean.

At the moment we *need* a boolean, because there's a binary choice: do we show the installer or not when you hit index.php.

I didn't want to use SETTINGS_PATH, because A) I want to be able to provide a settings file before installation, so an installer (or SAS hoster or ...) can provide some configuration for the installer to use. B) I think if we had good enough defaults, you could might legitimately run a site _without_ a settings file.

I was basically thinking of CIVICRM_INSTALLED as a "show the web installer?" settings - and it's kind of a tertiary:
TRUE => never show the installer; FALSE => show the installer; NULL = "AUTO" => determine whether to show the installer based on whether we can find an installed database. For the concerns above, AUTO isn't an ideal value.. but I can see it having uses.

From that perspective, I don't see the parallels with `set_include_path` / registering the classloader. It feels like legitimate config state to me.

ufundo replied:
> At the moment we need a boolean, because there's a binary choice: do we show the installer or not when you hit index.php.

This did make me think... what if we have a dedicated url path for the installer `/install` ? Then we can jump straight to `invoke` if we're not on that path. I still think we need a way to check whether *not* to show the installer on a site that is already installed, which I would see taking the same route

ufundo replied:
( For other UFs, wouldn't the ideal check would be "is_module_enabled('civicrm')" ? Though I can see that being complicated if the CiviCRM installation doesn't fit entirely within the "enable a module" paradigm of the UF. )

ufundo replied:
> I don't understand the idea of running without a settings file. The db credentials gotta go somewhere, right?

The hope is to provide with ENV vars. 

And I also think there's lots of legitimate use cases for having a settings file before install, in order to provide some settings to the installer. 

For both those reasons I think an explicit variable makes more sense than using CIVICRM_SETTINGS_PATH directly. 

> On the other side, "installation" refers to a process with various resources, phases, failure-points, future-features. Those are not the same size/complexity. Intuitively, there will be cases where the boolean is an anemic representation of the true installation-status.

Isn't this even more of a problem with `defined('CIVICRM_SETTINGS_PATH')`?

> Would renaming » CIVICRM_INSTALLED » CIVICRM_SHOW_INSTALLER be clearer?

 :shrug: I can see advantages either way but I think " the installer outputs a settings file which sets `CIVICRM_INSTALLED = 1`" makes more sense than "the installer has to unset CIVICRM_SHOW_INSTALLER".

ufundo committed:
remove old comment

Co-authored-by: Rich Lott <artfulrobot@users.noreply.github.com>

ufundo committed:
comment out ambition to check for installed database in WebEntrypoint

totten committed:
Drop CIVICRM_INSTALLED in favor of pre-existing CIVICRM_SETTINGS_PATH. Tidy up.

In the preceding commits of the PR, it updates `civicrm.settings.php` to add a new constant
CIVICRM_INSTALLED.  The constant is simply TRUE.  In other words,

> CIVICRM_INSTALLED is TRUE if-and-only-if civicrm.settings.php has been loaded

Of course, deploying new constants in `civicrm.settings.php` can be annoying.
Do we need to?  Well, there's already a similar constant...  sort of.

- Drupal, Backdrop, WordPress, and Joomla all set the constant CIVICRM_SETTINGS_PATH
  when they load `civicrm.settings.php`. You can check this variable to see
  if the settings are loaded.
- But Standalone doesn't currently. If we make it conform, then we don't
  need to add CIVICRM_INSTALLED.

(REF) WebEntryPoint - Simplify and privatize checkCiviInstalled

ufundo commented:
Have noticed this causes issues for the Standalone composer template https://github.com/civicrm/civicrm-standalone because that uses the StandaloneScaffold. 

Need a way to vary the paths in index.php / civicrm.standalone.php in that context. It should use the ones that are currently commented out in `index.php.txt` and `civicrm.standalone.php.txt`. 

Was wondering about the relative merits of:
a) have alternative templates ( `composer.index.php.txt`) and a flag to StandaloneScaffold to say "use composer variants)
b) use static files in the civicrm-standalone repository with the alternative versions
c) reinstate the "check in multiple places every time" type behaviour

ufundo committed:
use model civicrm.root path rather than userFrameworkResourceURL setting in installer

ufundo committed:
use civicrm.root url provided by asset plugin rather than hard code

ufundo committed:
cleanup standalone installer path setting

ufundo commented:
Have got it working with composer again - accompanying PR for the standalone template is https://github.com/civicrm/civicrm-standalone/pull/11

This required engaging with the path settings in the installer, so have added some "opinionated defaults" again, given the questions about the AppSettings approach to this.

ufundo commented:
> (Edit/aside: Oh, fun. I assumed CIVICRM_INSTALLED was entirely new -- at least, it doesn't exist on Civi-D7/D8+/BD/J. But on Civi-WP, it's a runtime-value that reflects the conditional file_exists(CIVICRM_SETTINGS_PATH) at the moment of initial load.)

Just noticed there's also [a SettingsManager setting `installed`](https://github.com/civicrm/civicrm-core/blob/2742fd477cf2e56dc9afdf04eff1962ba7e93801/CRM/Core/Config.php#L528) . Looks like a fun dynamic one. Though it sort of serves our purpose. :thinking:

ufundo commented:
Hi @artfulrobot

I hoped all the contentious stuff was in https://github.com/civicrm/civicrm-core/pull/30132 and that this one was nearly there. 

I think the most substantive outstanding issue here is `CIVICRM_INSTALLED` vs `CIVICRM_SHOW_INSTALLER`

@totten did you see my response on this above? 

I think if it will unblock then I will change to [if-settings-file-exists], but I don't love it...

ufundo committed:
fix whitespace

ufundo commented on Civi/Standalone/ErrorHandler.php:
```diff
@@ -0,0 +1,68 @@
+ */
+class ErrorHandler {
+
+  protected static bool $handlingError = FALSE;
```
oh yes good spot! :detective:

ufundo commented on Civi/Standalone/WebEntrypoint.php:
```diff
@@ -0,0 +1,113 @@
+
+    // the core folder is up two directories from this file
+    // TODO: use AppSettings to get the configured core path
+    $corePath = implode(DIRECTORY_SEPARATOR, array_slice(explode(DIRECTORY_SEPARATOR, __DIR__), 0, -2));
```
i like it

ufundo committed:
tidy as we go

Co-authored-by: Rich Lott <artfulrobot@users.noreply.github.com>

ufundo committed:
use class static rather than method static in ErrorHandler

ufundo committed:
need to find appRootPath if using cv

ufundo commented:
Thanks for re-reviewing and picking up all those bits @artfulrobot  :superhero: 

So it looks like your error was 2 things:

1. `civibuild` also needs a patch to be consistent with the tarball directory structure 
2. we can't rely on having the app root path from `civicrm.standalone.php` when using cv core:install

I think this fixes 1: https://github.com/civicrm/civicrm-buildkit/pull/860 (I renamed the variants in the process, to show the "dev" one is now basically the default, and what was "clean" is now a variant; also I remember in Ashbourne you saying you'd be very happy if someone found where that duplicate "upload" folder you always get on buildkit came from... and I think I found it)

For 2, I've pushed an additional commit here which I think fixes it for now. I don't love it because it's re-introducing another potential source of truth for what the directory structure is... but I can't see a quick way round it, and I think it's still generally an improvement.

ufundo commented:
Oh no... can I check is that with the cv patch as well @artfulrobot ? I feel like that error was what we used to get on that side. 

I'm pretty sure its working for me, but I' definitely finding it a bit tricky coordinating the patches. So I could have something slightly different

ufundo commented:
Acht. Sorry I see where I've dropped the "dont nix value passed from buildkit"

ufundo committed:
respect existing $model->cmsBaseUrl from buildkit

ufundo commented:
> Retested.

Thank you for sticking with it @artfulrobot  :pray: 

> ⚠️ Fatal error accessing the Directories settings page

I think I've seen this before, and looking at it now I think it is a general standalone bug. I can reproduce on clean buildkit install, `standalone-dev`  from master

> http://so1.localhost:8100/civicrm/dashlet/blog?reset=1&snippet=json&crmAngularModules=api4,crmDashboard,crmResource,crmUi,crmUtil,dialogService,jsonFormatter,ui.sortable

Think this is separate too, something Smarty related - I get `unknown modifier 'function_exists'`

ufundo commented:
Nice thanks :slightly_smiling_face: 

Directory settings crash a core bug, not even just a standalone one, so good to find! https://github.com/civicrm/civicrm-core/pull/30351

ufundo committed:
Standalone settings path not defined => *not* installed

totten committed:
Standalone Layout  - Resolve post-install warnings about path variables

If the `mandatorySettings` incorporate absolute paths or absolute URLs, then
there is a post-install status-alert telling you to use path-variables.

totten committed:
Standalone Layout - Use "l10n" and "ext" to match existing conventions

totten committed:
Standalone Layout - More evocative "upload(s)" folders.

For a moment, I really liked the symmetry of "public/uploads" and
"private/uploads".  And could be talked back into it.  It almost cleans up
the mess of "upload" terminology.

But I don't think it quite succeeds. The pre-existing mess is too much.

* The problem is "uploadDir": so there are two "upload(s)" folders, and
  neither is configured with "uploadDir".
* When you migrate from old-layout to new-layout, you'll have "upload(s)" in the
  before+after layouts... but the substance has to get swapped around.
* The programmatic default for `[civicrm.files]` and `[civicrm.private]`
  in the same place. If that happens, then the "uploads" would be simultaneously
  public+private. Obviously, we're discouraging that. But $wildInternet.

So I was looking for phrases that (a) would be more distinctive, (b) would have
some resonance. I think this does the job:

* "public/media" (images, videos, PDFs, etc that you're publishing;
  as in the CKEditor/KCFinder dialogs)
* "private/attachments" (files that you put onto a specific node;
  web UI calls these "Attachments" for activities and for civimail)

totten committed:
Standalone Layout - Simplify name of cache folder

totten commented:
@ufundo OK, so I was able to make both styles of builds. While trying them out, noticed a few things:

1. Post-install, some of the paths lead to "System Status" notices suggesting changes. (*Added some tweaks to fix those.*)
2. While testing the de-facto file-path for public-media, I needed to use https://github.com/civicrm/civicrm-packages/pull/408 along with custom `civicrm.config.php`.  We probably need 408 plus something else to address the `civicrm.config.php`.
3. Several folders had unexpected names. (At least, they differed from prior rounds of discussions.) Did an audit of the two builds and logged the layout in [the spreadsheet](https://docs.google.com/spreadsheets/d/15PwVHLvobd34gcQsKBaAnGQ_5_Fp99FnYnM5Mq7KhLs/edit#gid=1790058057) under "29960a". Used the sheet as a "swatch" for a mini-focus-group to elicit more feelings/opinions. Based on their feedback, landed on 29960c -- which preserves some changes (*e.g. dropping `data/`*); reverts some other changes (*e.g. restore `l10n/`*); and makes some new/different changes (e.g. `private/attachment/`). Pushed revisions to all three repos.
4. There's a pre-existing problem that CiviCRM's ecosystem is not consistent about whether to use plural form (e.g. `templates` vs `bin` vs `packages` vs `vendor`). Did an audit to see if plural was (empirically) more or less conventional:
    * Well, `civicrm-core` is the most mixed-up. Fun.
        * Pluralized: `bower_components` `packages` `release-notes` `settings` `templates` `tests` `tools`
        * Non-pluralized: `ang` `api` `bin` `css` `ext` `extern` `js` `i` `managed` `mixin` `schema` `sql` `vendor` `xml`
        * Not-applicable: `CRM` `Civi` `install` `distmaker` `l10n` `setup`
    * The pre-existing `files/civicrm/` tree (as seen on D7) doesn't use any plurals internally. But the parent (`files`) is plural.
    * Linux FHS is pretty consistent. (`bin` not `bins`; `log` not `logs`; `cache` not `caches`) with only one outlier (`/var/games`).
    * I think this is truest interpretation: all three precedents are mixed -- but all three precedents _lean toward_ singular.

At this point, we're MOP. In tandem with merging this, we should merge the parallel PRs:

* https://github.com/civicrm/civicrm-buildkit/pull/860
* https://github.com/civicrm/civicrm-standalone/pull/11

ufundo commented:
First off, great to have merged! 🎉

For the directory stuff, realising I didn't explain the changes I'd made here. I think originally they were in a subsequent branch but got bundled together to fix the composer stuff.


1. System Status notices
So I thought I had used tokenised versions to start with, but hit an issue that the tokens weren't being resolved during the install. Maybe something else has changed so it works now. Definitely prefer the tokens!

2. kcfinder
Can look into this. Looks like a sinilar situ if you want to use web cron, which at least one person did. One thing I'm confused about is how this works on D10 , which also doesn't have `civicrm.config.php` as far as I can see?
Thanks for the permissions piece of puzzle.

3. naming
I really like `media` and `attachment(s)`

I dont agree at all that `extensions` and `translations` are pointless changes. I think both are helpful to reduce overheads for new people/people who dont use Civi every day. It's one less thing to explain/remember/be confused by, and it makes people feel welcome. Particularly ext, given thats one of the first places lots of people will want to look. (I'm less clear if l10n is widely understood in the world of multi-lingual, but I've never seen it any other code context, and note D10 uses "translations")

4. plurals
I'd argue plurals similarly are more natural to most people. I'm not convinced that the conventions of Linux have much bearing for us; and especially given neither are conclusive, I'd say we should use what we think is best and try to push the dial in the right direction, rather than go with existing majority.
